File: go/api_keys.go
package realmid

import (
	ctxpkg "context"
	"encoding/json"
	"net/url"
	"time"
)

// APIKey is one entry from realm.APIKeys.* (SPEC §6.5). The struct is a
// union of the create-response and list-row wire shapes (issuer wins —
// see issuer swagger APIKey / APIKeyListItem):
//
//   - On create: ID, Value (the one-time secret), Scope, Label, ExpiresAt.
//   - On list:   ID, Prefix, Label, Role, CreatedAt, LastUsedAt, ExpiresAt,
//     RevokedAt are set.
//
// The issuer's create request uses `scope` while the list row reports
// `role`; the SDK surfaces both fields rather than papering over the
// asymmetry (logged for the issuer in sdk/TODO.md).
type APIKey struct {
	ID string `json:"id"`
	// Value is the raw secret key, returned ONLY on create (one-time reveal).
	Value string `json:"value,omitempty"`
	// Scope is echoed on create.
	Scope string `json:"scope,omitempty"`
	// Label is echoed on create AND returned on every list row (issuer
	// v0.61.0, ADR-085 §7). It is the only handle on a key: the plaintext is
	// never echoed and Prefix is derived from the stored hash, so an
	// rk_live_… found in a log cannot be traced to its row by value.
	Label string `json:"label,omitempty"`
	// Prefix is the non-secret key prefix (list rows), stable across logs.
	Prefix string `json:"prefix,omitempty"`
	// Role is the key's bound role as reported by the list endpoint.
	Role string `json:"role,omitempty"`
	// CreatedAt / LastUsedAt / RevokedAt are unix seconds (list rows).
	// LastUsedAt and RevokedAt are nil when the server omits them; a
	// non-nil RevokedAt means the key has been revoked.
	CreatedAt  int64  `json:"created_at,omitempty"`
	LastUsedAt *int64 `json:"last_used_at,omitempty"`
	RevokedAt  *int64 `json:"revoked_at,omitempty"`
	// ExpiresAt is the scheduled cutoff in unix seconds, nil for a
	// non-expiring key (ADR-085 §3). An expired key behaves exactly like a
	// revoked one at login and returns the same error envelope, so the two are
	// indistinguishable to a key holder — check Expired() and Revoked()
	// separately when you need to tell an operator which it was.
	ExpiresAt *int64 `json:"expires_at,omitempty"`
}

// Expired reports whether the key is past its scheduled cutoff as of now.
// A non-expiring key (ExpiresAt nil) is never expired.
func (k APIKey) Expired(now time.Time) bool {
	return k.ExpiresAt != nil && now.Unix() >= *k.ExpiresAt
}

// Revoked reports whether the key has been soft-deleted (revoked_at set).
func (k APIKey) Revoked() bool { return k.RevokedAt != nil }

// APIKeyCreate is the create payload. `Scope` is required; `Label` is an
// optional human-readable name.
type APIKeyCreate struct {
	Scope string `json:"scope"`
	Label string `json:"label,omitempty"`
	// TTLSeconds is the requested lifetime (ADR-085 §3). Omitting it AND
	// NonExpiring applies the issuer's built-in 90-day default. The floor is
	// 300s and a smaller value is REJECTED rather than clamped — clamping up
	// would hand back a key that outlives what was asked for.
	TTLSeconds int64 `json:"ttl_seconds,omitempty"`
	// NonExpiring requests a permanent key. A realm may hold at most one, and
	// at most 2 active platform keys in total (ADR-085 §2), so create can fail
	// with non_expiring_not_allowed or too_many_api_keys.
	NonExpiring bool `json:"non_expiring,omitempty"`
}

// APIKeysClient is realm.APIKeys.
type APIKeysClient struct {
	realm *Realm
}

func (c *APIKeysClient) Create(ctx ctxpkg.Context, body APIKeyCreate) (*APIKey, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var k APIKey
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/platforms/" + url.PathEscape(c.realm.realmID) + "/api-keys",
		Bearer: tok,
		Body:   body,
	}, &k); err != nil {
		return nil, err
	}
	return &k, nil
}

// List returns every API key associated with this realm. The issuer
// returns a paginated {items, next_cursor, total} envelope; we also
// accept a flat array or a legacy {api_keys} envelope for resilience.
func (c *APIKeysClient) List(ctx ctxpkg.Context) ([]APIKey, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var raw json.RawMessage
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET",
		Path:   "/platforms/" + url.PathEscape(c.realm.realmID) + "/api-keys",
		Bearer: tok,
	}, &raw); err != nil {
		return nil, err
	}
	return decodeAPIKeyList(raw)
}

func (c *APIKeysClient) Revoke(ctx ctxpkg.Context, id string) error {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return err
	}
	return c.realm.http.do(ctx, requestOptions{
		Method: "DELETE",
		Path:   "/platforms/" + url.PathEscape(c.realm.realmID) + "/api-keys/" + url.PathEscape(id),
		Bearer: tok,
	}, nil)
}

// decodeAPIKeyList tolerates the issuer's {items} envelope, a legacy
// {api_keys} envelope, or a flat array. Decoding straight into the typed
// APIKey struct picks up the wire field names (role/prefix/last_used_at)
// without manual key walking.
func decodeAPIKeyList(raw json.RawMessage) ([]APIKey, error) {
	var env struct {
		Items   []APIKey `json:"items"`
		APIKeys []APIKey `json:"api_keys"`
	}
	if err := json.Unmarshal(raw, &env); err == nil {
		if env.Items != nil {
			return env.Items, nil
		}
		if env.APIKeys != nil {
			return env.APIKeys, nil
		}
	}
	var arr []APIKey
	if err := json.Unmarshal(raw, &arr); err == nil {
		return arr, nil
	}
	return nil, &RealmError{Code: ErrCodeServerError, Message: "api-keys list: unexpected response shape"}
}

File: ts/src/api-keys.ts
/**
 * Realm API-key management — SPEC §6.5. Promoted to a top-level
 * namespace on the realm handle (`realm.apiKeys.*`).
 *
 * Wire shapes are issuer-authoritative ("code wins"): the create response
 * is the issuer `APIKey` (returns the one-time `value` secret), the list
 * rows are `APIKeyListItem`. The create request takes `scope` while the
 * list row reports `role` — the SDK surfaces both rather than papering
 * over the asymmetry (mirrors the Go SDK's `APIKey` struct). Field names
 * are snake_case on the wire, matching the rest of this SDK's DTOs.
 */

import type { HttpClient } from "./http.js";
import { RealmError } from "./errors.js";
import { paginate, readPage, type Paginated, type PageOpts } from "./pagination.js";

/** Create payload (issuer `APIKey` create). `scope` is required. */
export interface ApiKeyCreate {
  /** The role/scope the key is bound to. Required. */
  scope: string;
  /** Optional human-readable label shown in listings. */
  label?: string;
  /**
   * Requested lifetime (ADR-085 §3). Omitting this AND `non_expiring` applies
   * the issuer's built-in 90-day default. The floor is 300s and a smaller
   * value is REJECTED rather than clamped — clamping up would hand back a key
   * that outlives what was asked for.
   */
  ttl_seconds?: number;
  /**
   * Request a permanent key. A realm holds at most one non-expiring key and at
   * most 2 active platform keys in total (ADR-085 §2), so create can fail with
   * `non_expiring_not_allowed` (400) or `too_many_api_keys` (409).
   */
  non_expiring?: boolean;
}

/**
 * One API-key entry. A union of the create-response and list-row wire
 * shapes (issuer wins — see issuer swagger `APIKey` / `APIKeyListItem`):
 *
 *   - On create: `id`, `value` (the one-time secret), `scope`, `label`,
 *                `expires_at`.
 *   - On list:   `id`, `prefix`, `label`, `role`, `created_at`,
 *                `last_used_at`, `expires_at`, `revoked_at`.
 */
export interface ApiKey {
  id: string;
  /** Raw secret key — returned ONLY on create (one-time reveal). */
  value?: string;
  /** Echoed on create. */
  scope?: string;
  /**
   * The label supplied at create — echoed there and present on every list row
   * (issuer v0.61.0, ADR-085 §7). It is the ONLY handle on a key: the
   * plaintext is never echoed and `prefix` is derived from the stored hash, so
   * an `rk_live_…` found in a log cannot be traced to its row by value.
   */
  label?: string;
  /** Non-secret key prefix (list rows), stable across logs. */
  prefix?: string;
  /** The key's bound role as reported by the list endpoint (singular). */
  role?: string;
  /** Unix seconds (list rows). */
  created_at?: number;
  /** Unix seconds; null/absent until first use (list rows). */
  last_used_at?: number | null;
  /** Unix seconds; non-null means the key is revoked (list rows). */
  revoked_at?: number | null;
  /**
   * Unix seconds of the scheduled cutoff, or `null` for a non-expiring key
   * (ADR-085 §3). `null` is a VALUE, not an absence — "never expires" is a
   * fact the caller must be able to read. An expired key behaves exactly like
   * a revoked one at login and returns the same error envelope.
   */
  expires_at?: number | null;
}

export class ApiKeysClient {
  constructor(
    private readonly http: HttpClient,
    private readonly realmId: string,
  ) {}

  /**
   * Create a realm API key. Returns the row **plus** the one-time `value`
   * secret, which is shown only on creation and never returned by `list`.
   */
  async create(body: ApiKeyCreate): Promise<ApiKey> {
    return this.http.request<ApiKey>({
      method: "POST",
      path: `/platforms/${encodeURIComponent(this.realmId)}/api-keys`,
      body: {
        scope: body.scope,
        ...(body.label !== undefined ? { label: body.label } : {}),
      },
    });
  }

  /**
   * Paginate this realm's API keys.
   *
   * Returns the PAGER, not an array. The previous signature's own doc comment
   * already said the issuer answers a paginated
   * `{ items, next_cursor, total }` envelope — and then discarded it, so a
   * caller could neither page nor detect truncation. Read `.page().hasMore`,
   * or `for await` to walk everything.
   */
  list(opts?: PageOpts): Paginated<ApiKey> {
    const path = `/platforms/${encodeURIComponent(this.realmId)}/api-keys`;
    return paginate<ApiKey>(async (po) => {
      const raw = await this.http.request<unknown>({
        method: "GET",
        path,
        query: { cursor: po.cursor, limit: po.limit ?? opts?.limit },
      });
      return readPage<ApiKey>(raw);
    });
  }

  /** Soft-delete (sets `revoked_at`). */
  async revoke(id: string): Promise<void> {
    await this.http.request({
      method: "DELETE",
      path: `/platforms/${encodeURIComponent(this.realmId)}/api-keys/${encodeURIComponent(id)}`,
    });
  }
}

/**
 * Reports whether a key has been soft-deleted (`revoked_at` set).
 */
export function isApiKeyRevoked(k: ApiKey): boolean {
  return k.revoked_at != null;
}


